op-supervisor/supervisor/backend/db/db.go

```diff
@@ -0,0 +1,494 @@
+package db
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"slices"
+	"sync"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	entrySize                 = 24
+	searchCheckpointFrequency = 256
+
+	eventFlagIncrementLogIdx = byte(1)
+	//eventFlagHasExecutingMessage = byte(1) << 1
+)
+
+const (
+	typeSearchCheckpoint byte = iota
+	typeCanonicalHash
+	typeInitiatingEvent
+	typeExecutingLink
+	typeExecutingCheck
+)
+
+var (
+	ErrLogOutOfOrder  = errors.New("log out of order")
+	ErrDataCorruption = errors.New("data corruption")
+)
+
+type TruncatedHash []byte
+
+// dataAccess defines a minimal API required to manipulate the actual stored data.
+// It is a subset of the os.File API but could (theoretically) be satisfied by an in-memory implementation for testing.
+type dataAccess interface {
+	io.ReaderAt
+	io.Writer
+	io.Closer
+	Truncate(size int64) error
+}
```

Comment:
Maybe it makes sense to not couple the DB directly to a file reader-writer, but instead to an interface that can read/write entries by index? That way a lot of the IO read/write edge cases don't leak into the application-layer DB logic. And maybe with a method to read/write 2 entries at a time (for e.g. the checkpoint case).

Reply:
I think the write side is going to need to be adjusted a fair bit as part of fleshing out the error handling. I think the ideal case is to build up the entire set of entries to write in memory and write them with a single call - doesn't prevent failing part way through a write but will make a lot of things a lot simpler I think.

I like the idea of a simple entry db layer - it's already close to that given the `db.readEntry` and `db.writeEntry` calls but that would be worth splitting out and can then have a specific `truncateToEntry(idx)` method that Rewind uses.

Reply:
Introduced a basic entry db - it doesn't change much in the log db code currently but it should be a useful starting point when adding better error handling when writing. And it does pull a little bit of code out which is a win.  Haven't tried to support writing multiple entries at once yet - will leave that for the error handling work.